src/cata_ui.h

```diff
@@ -0,0 +1,429 @@
+        T *create_child( Args&&... args )
+        {
+            auto child = new T( std::forward<Args>(args)... );
+            children.push_back( child );
+            child->set_parent( this );
+            return child;
+        }
+};
+
+/**
+* @brief A simple text label
+*/
+class ui_label : public ui_element {
+    protected:
+        std::string text;
+
+        virtual void draw( WINDOW *win ) override;
+    public:
+        ui_label( std::string text, int x = 0, int y = 0, ui_anchor anchor = ui_anchor::top_left );
+
+        void set_text( std::string );
+
+        nc_color text_color = c_white;
+};
+
+/**
+* @brief A window with a border
+*/
+class bordered_window : public ui_window {
+    protected:
+        void draw( WINDOW *win ) override;
+    public:
+        bordered_window( size_t size_x, size_t size_y, int x = 0, int y = 0 );
+
+        using ui_window::draw;
+
+        nc_color border_color = BORDER_COLOR;
+};
+
+/**
+* @brief A window with tabs at the top.
+*
+### Creating tabs
+* Create a tab with the ```create_tab``` method.
+* It returns a ```ui_window *``` (Or something else if you pass it a template argument).
+* Use this window to let tabbed_window control visibility of elements you add to the window.
+*/
+class tabbed_window : public bordered_window {
+    private:
+        std::vector<std::pair<std::string, ui_element *>> tabs;
+        unsigned int tab_index = 0;
+    protected:
+        void draw( WINDOW *win ) override;
+    public:
+        tabbed_window( size_t size_x, size_t size_y, int x = 0, int y = 0 );
+
+        using ui_window::draw;
+
+        static constexpr int header_size = 3;
+
+        template<typename T = ui_window, typename... Ts>
+        T *create_tab( const std::string &tab, Ts&&... args )
+        {
+            auto tab_win = create_child<T>( get_rect().size_x - 2, get_rect().size_y - header_size - 1, 1, -1,
+                                            ui_anchor::bottom_left, std::forward<Ts>(args)... );
+            tabs.push_back({tab, tab_win});
+            tab_win->set_visible( tabs.size() == 1 );
+            return tab_win;
+        }
+
+        ui_event<> on_scroll;
+
+        void next_tab();
+        void previous_tab();
+        std::string current_tab() const;
+
+        input_broadcaster::listener default_action_handler();
+};
+
+auto default_draw_func = []( nc_color color, std::string txt )
+    {
+        return std::pair<std::string, nc_color>{txt, color}; // perform no transformation
+    };
+
+/**
+* @brief Basically, a list of text.
+*
+* One of the lines of text is highlighted (selected).
+* The list also has a scroll bar.
+*/
+template<typename D = std::string, typename Draw_Func = decltype(default_draw_func), D default_current = "">
+class ui_vertical_list : public ui_element {
+    private:
+        std::vector<D> items;
+        unsigned int scroll = 0;
+        unsigned int window_scroll = 0;
+
+        Draw_Func get_text_and_color;
```

Comment:
It might be better to use color tags. Tags are embedded in the text and print the text in different colors, e.g. `"this is in default color <color_red>this is in red</color>"`. You need to call different functions for printing, and have to add a second argument to `utf8_width` (`true` to indicate that it should skip those tags).

Color tags are commonly used at several places. It makes the interface a bit simpler: instead of a pair of text and color, you only need the text itself. It may also be a good idea to add support for color tags to the `ui_label` class. (This would only required changes in how it prints the text and how it determines the length).